app/Main.hs

```diff
@@ -65,5 +70,47 @@ numberWang = do
     putStrLn "Aah, would that it were NumberWang. Alas, it is not."
   numberWang
 
+toRevDigits :: Int -> [Int]
+toRevDigits n
+  | n <= 0    = []
+  | otherwise =  d : toRevDigits ds
+     where
+       d  = lastDigit n
+       ds = dropLastDigit n
+       lastDigit     = (`mod` 10)
+       dropLastDigit = (`div` 10)
+
+countInJapanese :: Int -> Maybe String
+countInJapanese x =
+  if searchNum x == Nothing
+    then reverse . drop 5 . reverse <$> genNumbers x
+  else searchNum x 
+
+genNumbers :: Int -> Maybe String
+genNumbers x = foldl look Nothing $ zip [0..] (toRevDigits x)
+    where
+      look :: Maybe String -> (Integer, Int) -> Maybe String
+      look acc (1, 1) = searchNum 10 <> Just " " <> acc      
```

Comment:
Can you please wrap these `String`, `Int` and `Integer` types with newtypes? Iit's hard to follow otherwise, especially with the presence of folds.